Source/qol/stash.cpp

```diff
@@ -494,53 +449,64 @@ bool UseStashItem(uint16_t c)
 
 	UseItem(MyPlayerId, item->_iMiscId, item->_iSpell);
 
-	if (Stash.stashList[c]._iMiscId == IMISC_MAPOFDOOM)
+	if (Stash.stashList[itemIndex]._iMiscId == IMISC_MAPOFDOOM)
 		return true;
-	if (Stash.stashList[c]._iMiscId == IMISC_NOTE) {
+	if (Stash.stashList[itemIndex]._iMiscId == IMISC_NOTE) {
 		InitQTextMsg(TEXT_BOOK9);
 		CloseInventory();
 		return true;
 	}
-	Stash.RemoveStashItem(c);
+	Stash.RemoveStashItem(item, itemPosition, itemIndex);
 
 	return true;
 }
 
-void StashStruct::RemoveStashItem(StashStruct::StashCell iv)
+void StashStruct::RemoveStashItem(Item *item, Point itemPosition, StashCell itemIndex)
 {
-	// Iterate through stashGrid and remove every reference to item
-	for (auto &row : Stash.GetCurrentGrid()) {
-		for (StashStruct::StashCell &itemId : row) {
-			if (itemId - 1 == iv) {
-				itemId = 0;
-			}
-		}
-	}
-
-	if (stashList.empty()) {
+	if (item == nullptr) {
 		return;
 	}
 
-	// If the item at the end of stash array isn't the one we removed, we need to swap its position in the array with the removed item
+	// get item's size to figure slot changes
+	const Size itemSize = GetInventorySize(*item);
+
+	// reset claimed slot IDs to zero
+	UpdateStashGrid(Stash.GetPage(), itemPosition, itemSize);
+
+	// not the last item, swap with last item to keep IDs contiguous
 	StashStruct::StashCell lastItemIndex = static_cast<StashStruct::StashCell>(stashList.size() - 1);
-	if (lastItemIndex != iv) {
-		stashList[iv] = stashList[lastItemIndex];
-
-		for (auto &pair : Stash.stashGrids) {
-			auto &grid = pair.second;
-			for (auto &row : grid) {
-				for (StashStruct::StashCell &itemId : row) {
-					if (itemId == lastItemIndex + 1) {
-						itemId = iv + 1;
-					}
-				}
-			}
-		}
+	if (lastItemIndex != itemIndex) {
+		Point swappedItemPosition = stashList[lastItemIndex].position;
+		const Size swappedItemSize = GetInventorySize(stashList[lastItemIndex]);
+		unsigned swappedItemPage = GetPageByItemId(lastItemIndex);
+		std::swap(stashList[itemIndex], stashList[lastItemIndex]);

```

Comment:
std::swap is unnecessary here, `stashList[itemIndex]` can be treated as if it's a garbage value at this point.

```suggestion
		stashList[itemIndex] = std::move(stashList[lastItemIndex]);
```